avoidance/src/transform_buffer.cpp

```diff
@@ -0,0 +1,128 @@
+namespace avoidance {
+
+TransformBuffer::TransformBuffer(float buffer_size_s) : buffer_size_(ros::Duration(buffer_size_s)) {
+  mutex_.reset(new std::mutex);
+};
+
+std::string TransformBuffer::get_key(const std::string& source_frame, const std::string& target_frame) const {
+  return source_frame + "_to_" + target_frame;
+}
+
+bool TransformBuffer::interpolateTransform(const tf::StampedTransform& tf_earlier, const tf::StampedTransform& tf_later,
+                                           tf::StampedTransform& transform) const {
+  // check if the requested timestamp lies between the two given transforms
+  if (transform.stamp_ > tf_later.stamp_ || transform.stamp_ < tf_earlier.stamp_) {
+    return false;
+  }
+
+  ros::Duration timeBetween = tf_later.stamp_ - tf_earlier.stamp_;
+  ros::Duration timeAfterEarlier = transform.stamp_ - tf_earlier.stamp_;
+  float tau = static_cast<float>(timeAfterEarlier.toNSec()) / timeBetween.toNSec();
+
+  tf::Vector3 translation = tf_earlier.getOrigin() * (1 - tau) + tf_later.getOrigin() * tau;
+  tf::Quaternion rotation = tf_earlier.getRotation().slerp(tf_later.getRotation(), tau);
+
+  transform.setOrigin(translation);
+  transform.setRotation(rotation);
+  return true;
+}
+
+bool TransformBuffer::isRegistered(const std::string& source_frame, const std::string& target_frame) const {
+  std::unordered_map<std::string, std::deque<tf::StampedTransform>>::const_iterator iterator =
+      buffer_.find(get_key(source_frame, target_frame));
+  if (iterator == buffer_.end()) {
+    return false;
+  } else {
+    return true;
+  }
+}
+
+void TransformBuffer::registerTransform(const std::string& source_frame, const std::string& target_frame) {
+  std::lock_guard<std::mutex> lck(*mutex_);
+  if (!isRegistered(source_frame, target_frame)) {
+    std::deque<tf::StampedTransform> empty_deque;
+    std::pair<std::string, std::string> transform_frames;
+    transform_frames.first = source_frame;
+    transform_frames.second = target_frame;
+    registered_transforms_.push_back(transform_frames);
+    buffer_[get_key(source_frame, target_frame)] = empty_deque;
+    ROS_WARN("transform buffer: Registered %s", get_key(source_frame, target_frame).c_str());
+  }
+}
+
+bool TransformBuffer::insertTransform(const std::string& source_frame, const std::string& target_frame,
+                                      tf::StampedTransform transform) {
+  std::lock_guard<std::mutex> lck(*mutex_);
+  std::unordered_map<std::string, std::deque<tf::StampedTransform>>::iterator iterator =
+      buffer_.find(get_key(source_frame, target_frame));
+  if (iterator == buffer_.end()) {
+    ROS_ERROR("TF cannot be written to buffer, unregistered transform");
+  } else {
+    // check if the given transform is newer than the last buffered one
+    bool new_tf = false;
+    if (iterator->second.size() == 0) {
+      new_tf = true;
+    } else if (iterator->second.back().stamp_ < transform.stamp_) {
+      new_tf = true;
+    }
+
+    if (new_tf) {
```

Comment:
```suggestion
    if (iterator->second.size() == 0 || iterator->second.back().stamp_ < transform.stamp_) {
```
And then get rid of the `new_tf` variable